lib/withOnyx.d.ts

```diff
@@ -0,0 +1,40 @@
+import {IsEqual} from 'type-fest';
+import {Key, CollectionKey, Selector, Value} from './types';
+
+type BaseMapping<TComponentProps, TOnyxProps> = {
+    canEvict?: boolean | ((props: Omit<TComponentProps, keyof TOnyxProps>) => boolean);
+    initWithStoredValues?: boolean;
+};
+
+type Mapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps, TOnyxKey extends Key | CollectionKey> = BaseMapping<TComponentProps, TOnyxProps> &
+    (
+        | (IsEqual<Value[TOnyxKey] | null, TOnyxProps[TOnyxProp]> extends true
+              ? {
+                    key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey);
+                }
+              : never)
+        | {
+              key: TOnyxKey;
+              selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
+          }
+        | {
+              key: (props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey;
+              selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
+          }
+    );
```

Comment:
@fabioh8010 could you explain what you're doing here?

Here is what I understand so far:

- For each `onxyProp`, we're creating a union type of `Mapping`s.
- We're evaluating `Mapping` for all Onyx keys
- I know we're doing that so that we can grab `TOnyxKey`
- I understand the following part
> ```ts
> IsEqual<Value[TOnyxKey] | null, TOnyxProps[TOnyxProp]> extends true
> ? {
>        key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey);
> }
> : never
> ```

- `IsEqual<Value[TOnyxKey] | null, TOnyxProps[TOnyxProp]>` helps to figure out which onyx key is mapped to the onyx prop
- I remember I realized a concern that the above method is not a reliable way to figure out onyx key. This is because the value type for the onyx prop can be different from the value type for the onyx key when the selector is used.

The part where I'm confused is the change you made to make `Mapping` a union type of different types of objects. I have three confusions:

- Now, `Mapping` is never `never` 😭 . That means that `OnyxPropMapping` is a union of bunch of unions for all possible onyx keys...?
- I feel like what you tried to accomplish here is to prevent `selector` to be passed when the type of onyx prop and onyx key is the same. Because we have a union type here, we actually cannot enforce that. And selector can be used even when the types of onyx prop and onyx key are the same (selector might just change the value of a property, not a type)

Reply:
@hayata-suenaga @blazejkustra 

As described in the PR description we have all this possible use cases for the mapping object of each Onyx prop:
- Onyx prop with `string` key
- Onyx prop with `string` collection key
- Onyx prop with `function` key
- Onyx prop with `function` collection key
- Onyx prop with `string` key and selector
- Onyx prop with `string` collection key and selector
- Onyx prop with `function` key and selector
- Onyx prop with `function` collection key and selector

The Mapping type is divided in order to catch specific use cases:
- Block 1
  ```ts
        | (IsEqual<Value[TOnyxKey] | null, TOnyxProps[TOnyxProp]> extends true
              ? {
                    key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey);
                }
              : never)
  ```
  - Satisfies `Onyx prop with string key`, `Onyx prop with function key`, `Onyx prop with string collection key` and `Onyx prop with function collection key` use cases.
  - As these use cases don't have selector the Onyx prop type must match with the Onyx value type, so we use `IsEqual<Value[TOnyxKey] | null, TOnyxProps[TOnyxProp]>` to make sure both types match.
- Block 2
  ```ts
        | {
              key: TOnyxKey;
              selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
          }
  ```
  - Satisfies `Onyx prop with string key and selector` and `Onyx prop with string collection key and selector` use cases.
  - As these use cases does have selector the Onyx prop type must match with the selector return type. Also, the Onyx value type must match with the selector parameter type.
- Block 3
  ```ts
        | {
              key: (props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey;
              selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
          }
  ```
  - Satisfies `Onyx prop with function key and selector` and `Onyx prop with function collection key and selector` use cases.
  - As these use cases does have selector the Onyx prop type must match with the selector return type. Also, the Onyx value type must match with the selector parameter type.

Blocks 2 and 3 are separated just to make the use cases more well-defined and readable.

Reply:
@fabioh8010 thank you for the detailed explanation. But I already understand what you explained in the previous message.

> Now, Mapping is never never 😭 . That means that OnyxPropMapping is a union of bunch of unions for all possible onyx keys...?

⬆️ This is my main concern. `OnyxPropMapping` is now a union of `Mapping` for all Onyx keys.

So this is possible:

```ts
type OnyxProps = {
  accountId: string;
};

type Props = OnyxProps & {
  foo: string;
};

type Result = typeof withOnyx<Props, OnyxProps>;
type MappingParam = Parameters<Result>[0];
type Mapping = MappingParam["accountId"];

const reportMapping: Mapping = {
  key: "isLoadingPaymentMethods",
  selector: (boolArg) => "randomString",
};
```

The above example uses your example. Now, `account` prop should be mapped to the accounts collection, but you can pass ANY key to the mapping object for `account`. In this example, `isLoadingPaymentMethods` is passed.

Reply:
Ohh I understand your question now @hayata-suenaga , but this situation is impossible to avoid. Actually it's not technically wrong as well.

If I'm using the `selector` what matters now is that `selector` return type must match `accountId` prop type. There is no way for TS to "understand" what specific key you need to use together with `selector` to supply your `accountId` prop because `accountId` type can refer to anything you want. Even the prop name could be different:

```ts
type OnyxProps = {
  someStringProp: string;
};

type Props = OnyxProps & {
  foo: string;
};

type Result = typeof withOnyx<Props, OnyxProps>;
type MappingParam = Parameters<Result>[0];
type Mapping = MappingParam["someStringProp"];

const reportMapping: Mapping = {
  key: "isLoadingPaymentMethods", // how TS will know what Onyx key is related to "someStringProp" prop?
  selector: (boolArg) => "randomString",
};
```

This enforcement you are talking about is only possible when not using the `selector`, because now I don't have data transformations in the mapping anymore and TS can enforce that the Onyx value type related to your `key` MUST match you prop type.

```ts
// ONYXKEYS.ts
type OnyxValues = {
    account: Account;
    ...
};
//////

type OnyxProps = {
  userAccount: Account | null;
};

type Props = OnyxProps & {
  foo: string;
};

type Result = typeof withOnyx<Props, OnyxProps>;
type MappingParam = Parameters<Result>[0];
type Mapping = MappingParam["userAccount"];

const reportMapping: Mapping = {
  key: "account", // I can only use a Onyx key which its Onyx value type matches "userAccount" prop type.
};
```

Reply:
With the current API, it's impossible to infer which Onyx key is mapped to each prop name.

But I also understands that we cannot manually pass Onxy Key type information to each mapping object (we cannot use type argument to a nested object).

The other workaround is passing another type argument to `withOnyx` which maps between prop name and Onxy key name...

Reply:
By the way, I asked a [question](https://expensify.slack.com/archives/C01GTK53T8Q/p1689799928393569) about `withOnyx` in the open source channel

Reply:
ah yea this is a compromise we have to make. But I think we should explain this in the code with extensive comments.

usually, TypeScript describes APIs correctly. If we have this kind of workaround solution that deviates from the correct API of implementation code, we should explain it really clearly.

but @fabioh8010 thank you so much for coming up with this solution. I think this is the best we can do and appreciate your effort really figuring out the best possible type safety 🙇